doc/source/api.rst

```diff
@@ -1,10 +1,259 @@
+turned into a regular Python object.
+
+Note that ``ray.get`` accepts either a single object ID or a list of object IDs.
+
+.. code-block:: python
+
+  import numpy as np
+
+  @ray.remote
+  def f():
+    return {'key1': ['value']}
+
+  # Get one object ID.
+  ray.get(f.remote())  # {'key1': ['value']}
+  # Get a list of object IDs.
+  ray.get([f.remote() for _ in range(2)])  # [{'key1': ['value']}, {'key1': ['value']}]
+
+Numpy arrays
+~~~~~~~~~~~~
+
+Numpy arrays are handled more efficiently than other data types, so **use numpy
+arrays whenever possible**.
+
+Any numpy arrays that are part of the serialized object will not be copied out
+of the object store. They will remain in the object store and the resulting
+deserialized object will simply have a pointer to the relevant place in the
+object store's memory.
+
+Since objects in the object store are immutable, this means that if you want to
+mutate a numpy array that was returned by a remote function, you will have to
+first copy it.
+
+.. autofunction:: ray.get
+
+Putting objects in the object store
+-----------------------------------
+
+The primary way that objects are placed in the object store is by being returned
+by a task. However, it is also possible to directly place objects in the object
+store using ``ray.put``.
+
+.. code-block:: python
+
+  x_id = ray.put(1)
+  ray.get(x_id)  # 1
+
+Note that ``ray.put`` is called under the hood in a couple situations.
+
+- It is called on the values returned by a task.
+- It is called on the arguments to a task, unless the arguments are simple
+  things like integers or short strings, lists, tuples, or dictionaries, in
+  which case they are serialized as part of the task specification.
+
+.. autofunction:: ray.put
+
+Waiting for a subset of tasks to finish
+---------------------------------------
+
+It is often desirable to adapt the computation being done based on when
+different tasks finish. For example, if a bunch of tasks each take a variable
+length of time, and their results can be processed in any order, then it makes
+sense to simply process the results in the order that they finish. In other
+settings, it makes sense to discard straggler tasks whose results may not be
+needed.
+
+To do this, we introduce the ``ray.wait`` primitive, which takes a list of
+object IDs and returns when a subset of them are available. By default it blocks
+until a single object is available, but the ``num_returns`` value can be
+specified to wait for a different number. If a ``timeout`` argument is passed
+in, it will block for at most that many milliseconds.
+
+The ``ray.wait`` function returns two lists. The first list is a list of object
+IDs of available objects (of length at most ``num_returns``), and the second
+list is a list of the remaining object IDs, so the concatenation of these two
+lists is equal to the list passed in to ``ray.wait`` (up to ordering).
+
+.. code-block:: python
+
+  import time
```

Comment:
I think the `ray.wait` examples might require more thought. It would be good to come up with an example that matches the use case you describe ("discard straggler tasks whose results may not be needed").

Also, we should give an example of why it might be nice for `ray.wait` to return object IDs instead of object values. More specifically, we should come up with an example where the processing for the ready IDs doesn't need the object values. Some kind of adaptive search could be a good example.